trenches2.py

```diff
@@ -66,17 +71,29 @@ def point_on_circle(center: dict, radius: float, radian: float) -> Tuple[float,
     y = center['y'] + (radius * math.sin(radian))
     return x, y
 
+def _LatLon_To_XY(Lat,Lon):
+  return P(Lat,Lon)
 
-def get_perpendicular_line(u_node: dict, v_node: dict, ref_point: dict) -> Tuple[dict, dict]:
-    dx = u_node['x'] - v_node['x']
-    dy = u_node['y'] - v_node['y']
-
-    # Perpendicular line
-    # TODO: flipping dx and dy is probably not correct but the point looks worse if we correct it
-    dx1 = -1 * dy
-    dy1 = dx
+def _XY_To_LatLon(x,y):
+  return P(x,y,inverse=True)
 
-    return {'x': ref_point['x'], 'y': ref_point['y']}, {'x': ref_point['x'] + dx1, 'y': ref_point['y'] + dy1}
+def get_perpendicular_line(u_node: dict, v_node: dict, ref_point: dict) -> Tuple[dict, dict]:
+    """
+    Returns the projection of a point over a line, corresponding to the perpendicular line.
+    :param u_node: The start node of the vector
+    :param v_node: The end node of the vector
+    :param ref_point: The point we would like to project
+    :return: The projected point over the vector
+    """
+    x1, y1 = _LatLon_To_XY(u_node['x'], u_node['y'])
+    x2, y2 = _LatLon_To_XY(v_node['x'], v_node['y'])
+    x3, y3 = _LatLon_To_XY(ref_point['x'], ref_point['y'])
+    if ((y2 - y1) ** 2 + (x2 - x1) ** 2) != 0:
+        k = ((y2 - y1) * (x3 - x1) - (x2 - x1) * (y3 - y1)) / ((y2 - y1) ** 2 + (x2 - x1) ** 2)
```

Comment:
your doing y2 - y1 and x2 - x1 three time. you can simplify this like
dy = y2 - y1
dx = x2 - 1